### PR TITLE
[FreeSurface] Fix linking error

### PR DESCRIPTION
**Pull request by joaquinirazabal**
I have just added an explicit instantiation trying to solve issue #6210. I do not have experience in this kind of problems nor have a windows machine to test if it is working correctly, so maybe it is not enough.

### CODE REVIEW

File: applications/FreeSurfaceApplication/custom_python/add_custom_io_to_python.cpp [hunk 1]
```diff
@@ -134,5 +134,7 @@ void  AddCustomIOToPython(pybind11::module& pymodule)
 }
 }  // namespace Python.
 
+template class GidIO<>;
```
Comment (roigcarlo) [inline applications/FreeSurfaceApplication/custom_python/add_custom_io_to_python.cpp]: I think he problem comes from the `GidIO<EdgebasedGidGaussPointsContainer,EdgebasedGidMeshContainer>`.

I don't have my win machine here right now, I will test it tomorrow

Comment (joaquinirazabal) [reply applications/FreeSurfaceApplication/custom_python/add_custom_io_to_python.cpp]: Hi @roigcarlo, I made more changes, maybe this are more in the way of solving the problem, in linux everything still works properly.

**Comment by roigcarlo** [comment]
@joaquinirazabal I am looking into this, its a little more complex of what I though initially. I take care of it.

**Comment by joaquinirazabal** [comment]
Ok @roigcarlo, thank you very much, anything you need, please, tell me.

**Comment by roigcarlo** [comment]
@KratosMultiphysics/technical-committee  agrees that the solution for this will be to implement GidIOBase using a singleton pattern.

**Comment by roigcarlo** [comment]
Linking #3637 and closing the old issue

**Comment by joaquinirazabal** [comment]
Hi, @roigcarlo the FreeSurfaceApplication is now not compiling also in ubuntu, I think because of this. Are you looking into it? Is there anything a can do to help? Thanks!

**Comment by roigcarlo** [comment]
Hi @joaquinirazabal 

The solution is to implement a singleton so the instance is the same across all the libraries that call that object. Unfortunately I still didn't have time to do it.

I will try to do it in the following days. Also, feel free to implement it yourself, if you do so I can take a look quicker.

**Comment by joaquinirazabal** [comment]
HI, @roigcarlo, I made some changes and now it is compiling correctly in ubuntu, however, although I tried to implement the GidIOBase class using a singleton pattern it already fails when compiled in windows...

**Comment by roigcarlo** [comment]
Hi @joaquinirazabal, this last couple of commits should make it compile in windows. I have no manner to test if the new custom GidIO of your app works, could yo give it a try?

Also, in order to make the changes I had to update the structure of the app so its split between core and interface (most of the apps do that anyway) hope you don't mind.

**Comment by roigcarlo** [comment]
Also ping @philbucher or @rubenzorrilla since I made changes to the core and I cannot auto-aprove this in case it works.

**Comment by joaquinirazabal** [comment]
Of course, do not mind about changing the structure, thanks @roigcarlo 
Now, there is another error :(
imagen

**Comment by roigcarlo** [comment]
That's odd. That variable no longer exists. 
Since the error is in the core and it is compiling in the CI I assume is a conflict with an old compilation 

Can you make a clean make of the core?

**Comment by joaquinirazabal** [comment]
That was the problem, now it is compiling in Windows without problem!

File: applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.cpp [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#include <custom_utilities/edgebased_gid_io.h>
+
+namespace Kratos {
+    template class GidIO<EdgebasedGidGaussPointsContainer,EdgebasedGidMeshContainer>;
```
Comment (rubenzorrilla) [inline applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.cpp:4]: I'd move the implementation of the functions in the header to the new cpp (maybe in a future PR to ease the review of this one).

File: applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h [hunk 3]
```diff
@@ -576,6 +568,8 @@ class EdgebasedGidMeshContainer
     const char* mMeshTitle;
 };//class EdgebasedGidMeshContainer
 
+KRATOS_API_EXTERN template class KRATOS_API(FREE_SURFACE_APPLICATION) GidIO<EdgebasedGidGaussPointsContainer,EdgebasedGidMeshContainer>;
```
Comment (rubenzorrilla) [inline applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h:571]: @roigcarlo @joaquinirazabal just for me to know,  can you clarify why this is needed? If you do the explicit instantiation in here why it is done again in the new cpp?

File: applications/FreeSurfaceApplication/tests/generalTests.py [hunk 4]
```diff
@@ -43,7 +43,7 @@ def ExecuteReservoirTests(self):
             self.checkResults()
 
     def setUp(self):
-        self.check_tolerance = 1e-6
+        self.check_tolerance = 1e-5
```
Comment (rubenzorrilla) [inline applications/FreeSurfaceApplication/tests/generalTests.py:46]: @joaquinirazabal may I ask why you required to change the tolerance? I think the changes in this PR must not affect the results of the tests.

File: kratos/includes/gid_io.h [hunk 5]
```diff
@@ -61,12 +61,28 @@ enum MultiFileFlag {SingleFile, MultipleFiles};
...
+public:
+    static GidIOBase& GetInstance();
+
+    int GetData();
+    void SetData(int data);
+
+private:
+    static void Create();
+
+    static GidIOBase* mpInstance;
```
Comment (rubenzorrilla) [inline kratos/includes/gid_io.h:85]: Please add some Doxygen documentation to these methods.

**Review by rubenzorrilla — COMMENTED** [review]
The implementation makes sense for me. 

Nonetheless, I'd like to confirm it in the next @KratosMultiphysics/technical-committee meeting as I have no expertise on this design pattern and this would affect almost everyone.

File: applications/FreeSurfaceApplication/tests/generalTests.py (hunk 4 above)
Comment (joaquinirazabal) [reply applications/FreeSurfaceApplication/tests/generalTests.py:46]: My guess is that other changes in the code (FluidDynamics or Meshing applications) are causing the tests to fail. However, the differences are very small, and, in my opinion, a tolerance 1e-5 for the DISTANCE function is more than enough.

Comment (rubenzorrilla) [reply applications/FreeSurfaceApplication/tests/generalTests.py:46]: Could be. I think it is reasonable too.

File: applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h (hunk 3 above)
Comment (roigcarlo) [reply applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h:571]: The one in the cpp is the explicit instantiation, the one in the h is the declaration with the classes from the FreeSurface. Both are needed as the base clase is templated and derived from the core

File: kratos/includes/gid_io.h (hunk 5 above)
Comment (roigcarlo) [reply kratos/includes/gid_io.h:85]: To be honest I don't know what they do, they were already there. I supose they control the number of uses? Not sure...

Comment (roigcarlo) [reply kratos/includes/gid_io.h:85]: I will add it to the singleton ones. Those I know for sure

File: applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FreeSurfaceApplication/custom_utilities/edgebased_gid_io.h:571]: Can you please copy&paste this as an inline comment in the code?

**Review by antonialarese — COMMENTED** [review]
Thanks @joaquinirazabal for taking care of all this (among other things!) and thanks to @roigcarlo and @rubenzorrilla for you help.

All the changes seems fine to me but I am not entitled to give any technical judgement on the implementation details!!

**Review by RiccardoRossi — APPROVED** [review]
@KratosMultiphysics/technical-committee agrees

**Pull request merged**